components/feature/session/src/main/java/mozilla/components/feature/session/SwipeRefreshFeature.kt

```diff
@@ -0,0 +1,68 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.session
+
+import android.view.View
+import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
+import mozilla.components.browser.session.SelectionAwareSessionObserver
+import mozilla.components.browser.session.Session
+import mozilla.components.browser.session.SessionManager
+import mozilla.components.concept.engine.EngineView
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+
+/**
+ * Feature implementation to add pull to refresh functionality to browsers.
+ *
+ * @param swipeRefreshLayout Reference to SwipeRefreshLayout that has an [EngineView] as its child.
+ */
+class SwipeRefreshFeature(
+    private val sessionManager: SessionManager,
+    private val swipeRefreshLayout: SwipeRefreshLayout
+) : SelectionAwareSessionObserver(sessionManager), LifecycleAwareFeature,
+    SwipeRefreshLayout.OnChildScrollUpCallback, SwipeRefreshLayout.OnRefreshListener {
+
+    init {
+        swipeRefreshLayout.setOnRefreshListener(this)
+        swipeRefreshLayout.setOnChildScrollUpCallback(this)
+    }
+
+    /**
+     * Start feature: Starts adding pull to refresh behavior for the active session.
+     */
+    override fun start() {
+        observeSelected()
+    }
+
+    /**
+     * Callback that checks whether it is possible for the child view to scroll up.
+     * If the child view cannot scroll up, attempted to scroll up triggers a refresh gesture.
+     */
+    override fun canChildScrollUp(parent: SwipeRefreshLayout, child: View?) =
+        if (child is EngineView) {
+            val result = child.canScrollVerticallyUp()
+            result
+        } else {
+            true
+        }
+
+    /**
+     * Called when a swipe gesture triggers a refresh.
+     */
+    override fun onRefresh() {
+        activeSession
+            ?.let { sessionManager.getEngineSession(it) }
```

Comment:
nit: `getEngineSession` without a parameter should already give you the engine session for the selected session (if there's one)

Reply:
Leaving this since `activeSession` may be the custom tab session now